pkg/boot/linux/load_linux_image.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package linux
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"syscall"
+
+	"github.com/u-root/u-root/pkg/boot/image"
+	"github.com/u-root/u-root/pkg/boot/kexec"
+	"github.com/u-root/u-root/pkg/dt"
+	"github.com/u-root/u-root/pkg/uio"
+	"golang.org/x/sys/unix"
+)
+
+type kimage struct {
+	segments kexec.Segments
+	entry    uintptr
+	cleanup  []func() error
+}
+
+func (k kimage) clean() {
+	for _, c := range k.cleanup {
+		if err := c(); err != nil {
+			Debug("Failure: %v", err)
+		}
+	}
+}
+
+const (
+	kernelAlignSize = 1 << 21 // 2 MB.
+)
+
+func mmap(f *os.File) (data []byte, ummap func() error, err error) {
+	s, err := f.Stat()
+	if err != nil {
+		return nil, nil, fmt.Errorf("stat error: %w", err)
+	}
+	if s.Size() == 0 {
+		return nil, nil, fmt.Errorf("cannot mmap zero-len file")
+	}
+	d, err := unix.Mmap(int(f.Fd()), 0, int(s.Size()), syscall.PROT_READ, syscall.MAP_PRIVATE)
+	if err != nil {
+		return nil, nil, fmt.Errorf("mmap failed: %w", err)
+	}
+
+	ummap = func() error {
+		if err := unix.Munmap(d); err != nil {
+			return fmt.Errorf("failed to unmap %s: %v", f.Name(), err)
+		}
+		return nil
+	}
+
+	return d, ummap, nil
+}
+
+// sanitizeFDT cleanups boot param properties from chosen node of the given FDT.
+func sanitizeFDT(fdt *dt.FDT) (*dt.Node, error) {
+	// Clear old entries in case we've already been through kexec to get
+	// to this instance of runtime.
+	chosen, _ := fdt.NodeByName("chosen")
+	if chosen == nil {
+		return nil, fmt.Errorf("no /chosen node in device tree")
+	}
+	for _, property := range []string{"linux,elfcorehdr", "linux,usable-memory-range", "kaslr-seed", "rng-seed", "linux,initrd-start", "linux,initrd-end"} {
+		chosen.RemoveProperty(property)
+	}
+
+	return chosen, nil
+}
+
+func kexecLoadImage(kernel, ramfs *os.File, cmdline string, opts KexecOptions) (*kimage, error) {
+	fdt, err := dt.LoadFDT(opts.DTB)
+	if err != nil {
+		return nil, fmt.Errorf("loadFDT(%s) = %v", opts.DTB, err)
```

Comment:
I may be missing something but ISTM a lot of these %v could be %w, in case we can get to tests later.